### PR TITLE
feat!: replaces callbacks with promises for confirmed/acknowledged service methods

### LINKED ISSUE #50: fix: per-client timeout for confirmed services
I've often found timeouts to be a source of pain when the event loop gets very busy due to high rates of network I/O and, therefore, I'm a little wary of the timeout at https://github.com/bacnet-js/client/blob/6ed62999d1559d313bc06b82d1c3a017442c5fb3/src/lib/client.ts#L256, which is instantiated on a per-request basis.

Changing the values of `_invokeStore` to the shape 

```typescript
{ 
    deferred: Deferred<any>; // can be a callback if #49 maintains support for both
    createdAt: number; // milliseconds from epoch, i.e. Date.now()
 }
```

and changing the cleanup function to loop upon `_invokeStore` and only resolve entries with `createdAt` older than a configurable amount of milliseconds should allow us to have a single timeout which would only need to be re-instantiated when needed:

```typescript
let timeout: NodeJS.Timeout | null = null;

const TIMEOUT_MILLIS = 1_000;

const clearOutTimedOutRequests = () => {
  timeout = null;
  const now = Date.now();
  for (let key in this._invokeStore) {
    if (this._invokeStore[key].createdAt < now - TIMEOUT_MILLIS) {
      this._invokeStore[key].deferred.reject(new Error('timed out'));
      delete this._invokeStore[key];
    }
  }
  if (Object.keys(this._invokeStore).length > 0) {
    setTimeoutIfNotPresent();
  }
};

const setTimeoutIfNotPresent = () => {
  if (timeout === null) {
    setTimeout(clearOutTimedOutRequests, TIMEOUT_MILLIS)
  }
};

// then use setTimeoutIfNotPresent() on every request.
```

I haven't actually stumbled into an issue with timeouts, yet, but thought I'd leave this here before I forgot.

### LINKED ISSUE #51: feat: drop callback based methods in favor of async ones
See suggestiong by @jacoscaz https://github.com/bacnet-js/client/pull/49#issuecomment-3051393914

### PR DESCRIPTION
**Pull request by jacoscaz**
- Closes #50 (WIP)
- Closes #51 
- Fixes https://github.com/bacnet-js/client/issues/51#issuecomment-3052395150

BREAKING CHANGE: this PR changes the consumer-facing API of this library, replacing callbacks with promises

### CODE REVIEW

**Comment by jacoscaz** [comment]
This is still a draft / WIP but feedback would be greatly appreciated.

File: src/lib/client.ts [hunk 1]
```diff
@@ -168,6 +163,27 @@ const confirmedServiceMap: BACnetEventsMap = {
...
+
+	promise: Promise<T>
+
+	constructor() {
+		this.promise = new Promise((resolve, reject) => {
+			this.resolve = resolve
+			this.reject = reject
+		})
+	}
+}
```
Comment (robertsLando) [inline src/lib/client.ts:178]: IMO we could move all invokeStore logic into a separated class so we cloud remove some internal client methods there. WDYT?

Comment (robertsLando) [reply src/lib/client.ts:178]: Of course also into a separeted file

File: src/lib/client.ts [hunk 2]
```diff
@@ -256,40 +254,50 @@ export default class BACnetClient extends TypedEventEmitter<BACnetClientEvents>
...
 
+	private _addDeferred(id: number): Promise<NetworkOpResult> {
+		const deferred = new Deferred<NetworkOpResult>()
+		this._invokeStore[id] = deferred
+		trace(
+			`InvokeId ${id} callback added -> timeout set to ${this._settings.apduTimeout}.`, // Stack: ${new Error().stack}`,
+		)
 		const timeout = setTimeout(
-			toCall.bind(this, new Error('ERR_TIMEOUT')),
+			deferred.reject.bind(null, new Error('ERR_TIMEOUT')),
```
Comment (robertsLando) [inline src/lib/client.ts]: when timeout triggers could it happen that for some reason the deferred promise resolves? Do we have a guard for this? I usually override `resolve/reject` functions to handle this

File: src/lib/client.ts [hunk 3]
```diff
@@ -1169,44 +1131,26 @@ export default class BACnetClient extends TypedEventEmitter<BACnetClientEvents>
...
-		callback: ErrorCallback,
-	): void
-	writePropertyMultiple(
-		address: BACNetAddress,
-		values: WritePropertyMultipleObject[],
-		options: ServiceOptions,
-		callback: ErrorCallback,
-	): void
-	writePropertyMultiple(
+	async writePropertyMultiple(
```
Comment (robertsLando) [inline src/lib/client.ts:1174]: I see you removed the signature that allowed to call this without options, I would add that and make options optional

File: src/lib/client.ts [hunk 4]
```diff
@@ -1620,28 +1527,21 @@ export default class BACnetClient extends TypedEventEmitter<BACnetClientEvents>
...
-			next()
-		})
+		await this._addDeferred(settings.invokeId)
 	}
 
 	/**
 	 * Subscribes to Change of Value (COV) notifications for a specific property
 	 */
-	public subscribeProperty(
+	public async subscribeProperty(
```
Comment (robertsLando) [inline src/lib/client.ts:1561]: I think options may be optional here too and maybe on all other methods as well

**Review by robertsLando — CHANGES_REQUESTED** [review]
Well done so far! :)

File: src/lib/client.ts (hunk 1 above)
Comment (jacoscaz) [reply src/lib/client.ts:178]: Haha, you read my mind. I had come to the same conclusion. Will do!

**Comment by jacoscaz** [comment]
Thanks for the super-quick feedback! I have a bunch of stuff I need to sort out today but will come back to this later on tonight.

File: src/lib/client.ts (hunk 1 above)
Comment (jacoscaz) [reply src/lib/client.ts]: Done in https://github.com/bacnet-js/client/pull/52/commits/f900d501850d6344dbec908fd11e45d0077fe094

File: src/lib/client.ts (hunk 2 above)
Comment (jacoscaz) [reply src/lib/client.ts]: I don't _think_ this is possible as of the latest commit that refactored everything into the dedicated `InvokeStore` class but I'm also not 100% sure. Will check later in the day.

**Review by robertsLando — CHANGES_REQUESTED** [review]
Ok by thinking at this again I see a good reason to call clear in client close method. Actually I also think this is an issues as we are not clearing invokestore in client close.

Also I don't know if there is a way to make store async as well and actually make it return when the transport is gracefully closed

File: src/lib/invokestore.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import debugLib from 'debug'
+
+import { type NetworkOpResult } from './types'
+import { Deferred } from './utils'
+
+const debug = debugLib('bacnet:client:debug')
+const trace = debugLib('bacnet:client:trace')
+
+export class InvokeStore {
```
[the remark this answers is not in the document]
Comment (jacoscaz) [reply src/lib/invokestore.ts]: done in https://github.com/bacnet-js/client/pull/52/commits/8b426551a4e9fffb17c11dad317f9da2bf8f159f

File: src/lib/utils.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+	resolve!: (value: T) => void
+	reject!: (err: Error) => void
+	promise: Promise<T>
+	constructor() {
+		this.promise = new Promise((resolve, reject) => {
+			this.resolve = resolve
+			this.reject = reject
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (jacoscaz) [reply src/lib/utils.ts]: Ah yes, I had planned on doing this but forgot. Thank you for reminding me! I'm quite happy with the current implementation as it hides all the deferring logic from consumers. Done in  https://github.com/bacnet-js/client/pull/52/commits/fa4a3edd8b4294c4379d195c15a2eac527d26637 .

File: src/lib/invokestore.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+				entry.deferred.resolve(result)
+			}
+		} else {
+			debug('InvokeId', id, 'not found -> drop package')
+			trace(`Stored invokeId: ${Array.from(this.#entries.keys())}`)
+		}
+	}
+
+	#clear = () => {
+		this.#clearTimeout = null
```
[the remark this answers is not in the document]
Comment (jacoscaz) [reply src/lib/invokestore.ts]: done in https://github.com/bacnet-js/client/pull/52/commits/4de13a12732d22d66c79b521d127506c59e67c0b

**Comment by jacoscaz** [comment]
> Ok by thinking at this again I see a good reason to call clear in client close method. Actually I also think this is an issues as we are not clearing invokestore in client close.

First pass at this in https://github.com/bacnet-js/client/pull/52/commits/0e3eb2364df014e418b6a510ab2e422b37c0ab52
 
> Also I don't know if there is a way to make store async as well and actually make it return when the transport is gracefully closed

We likely need the transport to emit a `close` event and listen to that from the client, which can then call its own `close()` method. If so, I think this would fall within the scope of a separate PR as it needs some work on transports first.

File: src/lib/invokestore.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		const now = Date.now()
+		this.#entries.forEach(({ deferred, createdAt }, id) => {
+			if (now - createdAt > this.#timeout) {
+				deferred.reject(new Error('ERR_TIMEOUT'))
+				this.#entries.delete(id)
+			}
+		})
+		if (this.#entries.size > 0) {
+			this.#scheduleClear()
+		}
```
[the remark this answers is not in the document]
Comment (jacoscaz) [reply src/lib/invokestore.ts]: Done in https://github.com/bacnet-js/client/pull/52/commits/9df5204f008eb19a8181bf7826977d33d3f06c54

**Comment by robertsLando** [comment]
I have created https://github.com/bacnet-js/client/issues/53 to keep track of that

File: src/lib/requestmanager.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	constructor(timeout: number) {
+		this.#entries = new Map()
+		this.#timeout = timeout
+		this.#clearTimeout = null
+	}
+
+	add(id: number): Promise<NetworkOpResult> {
+		const deferred = new Deferred<NetworkOpResult>()
+		this.#entries.set(id, { deferred, createdAt: Date.now() })
+		this.#scheduleClear()
```
[the remark this answers is not in the document]
Comment (robertsLando) [reply src/lib/requestmanager.ts:37]: If you still want to keep a single timeout btw you should not reschedule it on add, instead you should keep it running and then reschedule based on the createdAt of next request

[the remark this answers is not in the document]
Comment (robertsLando) [reply src/lib/requestmanager.ts:37]: Let me know if I explained it well otherwise I will be more specific, this will require some refactor of requestmanager code

[the remark this answers is not in the document]
Comment (jacoscaz) [reply src/lib/requestmanager.ts:37]: Mh... `scheduleClear()` does not override an existing timeout. So, the moment we get a burst of requests, the first one will result in a first timeout after `apduTimeout` milliseconds. After that, the `scheduleTimeout()` call at the end of `clear()` guarantees that, until we run out of requests, we'll get a new timeout every `apduTimeout` milliseconds.

[the remark this answers is not in the document]
Comment (robertsLando) [reply src/lib/requestmanager.ts:37]: > scheduleClear() does not override an existing timeout

Yeah sorry you right I'm blind 😆  btw the second part IMO is still valid:

Example timeout 3 seconds

1. First request comes, timeout set
2. Second request comes after 2 seconds. Timeout is untuched
3. Timeout triggers, cancels first request and rescheduled one

At this point the next timeout will trigger after 3 more seconds resulting in second request to timeout after 5 seconds instead of 3. In order to prevent this and ensure that timeouts are respected for each request we should reschedule the timeout based on the next request in queue created at

[the remark this answers is not in the document]
Comment (jacoscaz) [reply src/lib/requestmanager.ts:60]: I was a little skeptical at first because I was afraid that having an array around would have implied having O(n) operations in hot code paths (specifically `add()` and `resolve()`). However, I think I have found a good compromise in https://github.com/bacnet-js/client/pull/52/commits/b4af8fa1428a8719c21843407265b55a61a105e0 .

File: src/lib/requestmanager.ts [hunk 10]
```diff
@@ -17,30 +18,52 @@ class Deferred<T> {
...
 		this.#scheduleClear()
 		trace(
-			`InvokeId ${id} callback added -> timeout set to ${this.#timeout}.`, // Stack: ${new Error().stack}`,
+			`InvokeId ${invokeId} callback added -> timeout set to ${this.#timeout}.`, // Stack: ${new Error().stack}`,
 		)
 		return deferred.promise.finally(() => {
-			debug(`InvokeId ${id} deferred called`)
-			this.#entries.delete(id)
+			debug(`InvokeId ${invokeId} deferred called`)
+			this.#requestsById.delete(invokeId)
```
Comment (robertsLando) [inline src/lib/requestmanager.ts:66]: seems you are deleting requestsById both here and below, I think you can keep this and remove the one below. Also you are never deleting requests from requestsByTime. I know you are filtering them later but I'm wondering if it could make sense to delete them here as well to prevent having the two structures unsynced?

File: src/lib/requestmanager.ts [hunk 11]
```diff
@@ -51,18 +74,18 @@ export class RequestManager {
 		err: Error | null | undefined,
 		result?: NetworkOpResult,
 	) {
-		const entry = this.#entries.get(id)
-		if (entry) {
+		const request = this.#requestsById.get(id)
+		if (request) {
 			trace(`InvokeId ${id} found -> call callback`)
-			this.#entries.delete(id)
+			this.#requestsById.delete(id)
```
Comment (robertsLando) [inline src/lib/requestmanager.ts:80]: this is the other place where you delete the requestById

File: src/lib/requestmanager.ts [hunk 12]
```diff
@@ -72,23 +95,42 @@ export class RequestManager {
...
+			// Request is still pending
+			return true
 		})
-		debug(`Cleared ${qty - this.#entries.size} entries.`)
-		debug(`There are ${this.#entries.size} entries pending.`)
-		if (this.#entries.size > 0) {
+		assert(
+			this.#requestsById.size === this.#requestsByTime.length,
+			`Index size mismatch  ${this.#requestsById.size} !== ${this.#requestsByTime.length}`,
+		)
```
Comment (robertsLando) [inline src/lib/requestmanager.ts:117]: wondering if this is a check you added for tests, or you really want to have this as watchguard?

File: src/lib/requestmanager.ts [hunk 13]
```diff
@@ -72,23 +95,42 @@ export class RequestManager {
...
 	#scheduleClear() {
-		if (this.#clearTimeout === null) {
-			this.#clearTimeout = setTimeout(this.clear, this.#timeout)
+		if (this.#clearTimeout === null && this.#requestsByTime.length > 0) {
+			// setTimeout() effectively becomes like setImmediate() if the delay value is 0 or negative
+			this.#clearTimeout = setTimeout(
+				this.clear,
+				// We add 100 milliseconds to ensure that we can never schedule more than
+				// 10 timeouts per second even in worst-case bursts scenarios
+				this.#requestsByTime[0].expiresAt - Date.now() + 100,
```
Comment (robertsLando) [inline src/lib/requestmanager.ts:132]: I would firstly store this to a var, do some checks on them for example ensure it is grater then 0, print a log about the new timeout set and the set the timeout

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by robertsLando** [comment]
Last but not least, I would add some unit tests for the new request manager, I suggest to use mock timers: https://nodejs.org/api/test.html#timers

File: src/lib/requestmanager.ts (hunk 10 above)
Comment (jacoscaz) [reply src/lib/requestmanager.ts:66]: > seems you are deleting requestsById both here and below, I think you can keep this and remove the one below.

Ah, yes, well spotted! I deleted the one below as this is the only one that is triggered on both resolution and timeouts. Done in https://github.com/bacnet-js/client/pull/52/commits/1079027100e655907586130d21883785fb6e73ec . 

> Also you are never deleting requests from requestsByTime. I know you are filtering them later but I'm wondering if it could make sense to delete them here as well to prevent having the two structures unsynced?

In principle I try to keep O(n) operations out of hot code paths. I do realise that, in practice, this might never become a problem but, when possible, I always find it beneficial for future me to keep things as theoretically sound as possible.

Also, the two structures are never completely unsynced. Rather, the array can, at times, be a superset of the map index.

File: src/lib/requestmanager.ts (hunk 11 above)
Comment (jacoscaz) [reply src/lib/requestmanager.ts]: Done in https://github.com/bacnet-js/client/commit/1079027100e655907586130d21883785fb6e73ec

File: src/lib/requestmanager.ts (hunk 13 above)
Comment (jacoscaz) [reply src/lib/requestmanager.ts]: Done in https://github.com/bacnet-js/client/pull/52/commits/13f2b1dc9f0f5ea921e6c58c070a0f1e469a2630

**Comment by jacoscaz** [comment]
> Last but not least, I would add some unit tests for the new request manager, I suggest to use mock timers: https://nodejs.org/api/test.html#timers (quite sure copilot can do a nice job here If you describe well the edge cases you want to cover)

Yes, absolutely. Unit tests are on my todo list but I wanted to finalise the inner workings of the class before spending time on tests.

File: src/lib/requestmanager.ts (hunk 12 above)
Comment (robertsLando) [reply src/lib/requestmanager.ts:116]: I think we should be 100% sure we don't leak anything instead of causing unexpected client exceptions like this

**Review by robertsLando — COMMENTED** [review]
I see you named spec file with request-manager, I would use the `-` separator then also in the src/lib one

There are lint issues on tests

About tests: I really suggest to use https://nodejs.org/api/test.html#timers to create some use case tests

**Comment by jacoscaz** [comment]
> no clue if could be worth giving a try to sinon fake timers then

We could think about it if the current approach proves insufficient but, IMHO, the test suite is looking decent and has ample room for growth even with actual timers. Yes, the mock timers API may allow us to shave a few seconds off but not to a degree that would make any practical difference. Wrapping `setTimeout` with `mock.fn()` is what actually provides most of the value and that already works well.

File: src/lib/request-manager.ts [hunk 14]
```diff
@@ -38,16 +38,27 @@ export class RequestManager {
 	/** Array of requests ordered by creation time */
 	#requestsByTime: RequestEntry[]
 
-	#timeout: number
-	#clearTimeout: NodeJS.Timeout | null
+	/** Minimum time to wait between scheduled clearings of pending requests */
+	#clearTimeoutDelay: number
 
+	/** Id of the timeout for the next scheduled clearing of pending requests */
+	#clearTimeoutId: NodeJS.Timeout | null
```
[the remark this answers is not in the document]
Comment (jacoscaz) [reply src/lib/request-manager.ts]: Done in https://github.com/bacnet-js/client/pull/52/commits/0a1daa0532892215d669e810b217c9c10f3c8bd5

**Review by robertsLando — COMMENTED** [review]
LGTM. Well done 🙏🏼

**Review by robertsLando — APPROVED** [review]
